messaging/src/main/java/org/axonframework/messaging/NamespaceMessageTypeResolver.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2010-2025. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.messaging;
+
+import jakarta.annotation.Nonnull;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Implementation of the {@link MessageTypeResolver} that maintains a mapping of payload types to their corresponding
+ * {@link MessageType}s. This resolver organizes message types under a common namespace.
+ * <p>
+ * The resolver will throw a {@link MessageTypeNotResolvedException} when encountering a payload type for which no
+ * mapping has been defined. If fallback behavior is desired, use the
+ * {@link NamespaceMessageTypeResolverBuilder#fallback(MessageTypeResolver)} method during configuration to provide an
+ * alternative resolver.
+ * <p>
+ * Use the {@link #namespace(String)} static method to start building a resolver with a default namespace.
+ *
+ * @author Mateusz Nowak
+ * @since 5.0.0
+ */
+public class NamespaceMessageTypeResolver implements MessageTypeResolver {
+
+    private final Map<Class<?>, MessageType> mappings;
+
+    private NamespaceMessageTypeResolver(
+            @Nonnull Map<Class<?>, MessageType> mappings
+    ) {
+        Objects.requireNonNull(mappings, "Mappings may not be null");
+        this.mappings = mappings;
+    }
+
+    /**
+     * Sets a new namespace for subsequent message mappings.
+     *
+     * @param namespace The namespace to use for message types created after this call.
+     * @return The builder instance for method chaining.
+     */
+    public static NamespaceMessageTypeResolverBuilder namespace(@Nonnull String namespace) {
+        return new InternalNamespaceMessageTypeResolverBuilder(namespace, Map.of(), null);
+    }
+
+    @Override
+    public MessageType resolve(Class<?> payloadType) {
+        var messageType = mappings.get(payloadType);
+        if (messageType == null) {
+            throw new MessageTypeNotResolvedException(
+                    "No MessageType found for payload type [" + payloadType.getName() + "]");
```

Comment:
This exception is a) directed to the user or b) used as control flow by the FallbackMessageTypeResolver.

While this works, exceptions in Java are very expensive. And this is bad, especially for an operation that will happen many times. 

Maybe the FallbackMessageTypeResolver can remember which classes threw an exception? That way, the exception-as-control-flow is only used once per class

Reply:
The idea it's for both - a and b, also to inform user what has happened, but only if there is no fallback.
I've removed the "fallback" logic to dedicated class, assuming it's something general, and I wanted the composition, whereas now it's more expensive indeed. Before it was incorporated in the `NamespaceMessageTypeResolver` itself. 

I don't expect it to happen so often - the best practice should be to define all your messages.
But I agree - it's kind of the optimization if there is a fallback, I know I should use it without throwing the exception. If it's what you had on your mind I can do it.

Reply:
Instead of throwing an exception for unresolved exceptions, could we return an Optional?

Reply:
Wouldn't hurt to switch it to an `Optional` instead. That will mean all places we use the `MessageTypeResolver` will change, as they need to deal with the `Optional` at that point; and perhaps throw an exception themselves. If you'd go this route, @MateuszNaKodach, feel free to change the return type of the `MessageTypeResolver` in another PR.

Reply:
I kept throwing the exception to the responsibility of the interface. 
So the interface support both usage styles. What do you think? Then we don't need to handle it and throw exception on our own in many places. It's widely used in Kotlin, such methods like `getOrThrow` or even in Optional API `orElseThrow`.

Reply:
I didn't notice "another PR" and I've done it just here. 
What do you think about this solution?